### ISSUE TITLE
Emulate empty floppy drive and CD-ROM drive

### ISSUE THREAD

**Issue by frank-deng** [joncampbell123/dosbox-x#4853]
### What you want

Is it possible to emulate a scenario where Floppy/CD-ROM is present but without floppy disk/CD-ROM disc inserted, like VirtualBox does.

### Have you checked that no similar feature request(s) exist?

- [X] I have searched and didn't find any similar feature request.

### Code of Conduct & Contributing Guidelines

- [X] I agree to follow the code of conduct and the contributing guidelines.

**Comment by rderooy** [comment]
Yes, there should be an option for `imgmount` to "mount" an empty drive. Especially for running Win9x this would be helpful. The work-around for now is to always have some media in the drive, which for a floppy drive can be an empty disk.

For floppy, this will also require setting the floppy drive type.

**Comment by joncampbell123** [comment]
Yes! Not everybody 1990s kid booted their computers with a floppy and a CD inserted in the drives. :)

**Comment by joncampbell123** [comment]
I got an initial implementation for mounting an empty floppy and empty CD-ROM drive finished.

Latest commit:

IMGMOUNT 0 -fs none -t floppy -empty
IMGMOUNT e: -t iso -empty

**Comment by rderooy** [comment]
@joncampbell123 nice!

But you should probably add a flag to set the floppy BIOS type, otherwise some software will get confused. I'm thinking of Win9x thinking you have a 5.25" drive (type 2), while you are wanting to use 3.5" 1.44MB (type 4).

**Comment by joncampbell123** [comment]
> @joncampbell123 nice!
> 
> But you should probably add a flag to set the floppy BIOS type, otherwise some software will get confused. I'm thinking of Win9x thinking you have a 5.25" drive (type 2), while you are wanting to use 3.5" 1.44MB (type 4).

It's going to have to be a 1.44MB drive for now, but I'll address that after release March 1st.

**Comment by joncampbell123** [comment]
Actually, I added support for the -size numbers for the empty drive, and it will try to match BIOS floppy type code too. That should complete this request.

**Comment by rderooy** [comment]
I'm not sure what the issue is here, but compiled the latest code and cannot seem to mount an empty floppy or CD-ROM disk drive.
some things I tried:

![image](https://github.com/joncampbell123/dosbox-x/assets/17701098/eb1ac990-8789-4214-a73c-181f64c0b846)

**Comment by joncampbell123** [comment]
Syntax changed.

It is now:

IMGMOUNT 0 empty -t floppy -fs none

and

IMGMOUNT e: empty -t cdrom

The file name "empty" is now handled as a special name that means "empty drive".

**Comment by rderooy** [comment]
Thanks,

That works for a 3.5" and CD-ROM drive, it does not seem to work for a 5.25". Or at least, after booting Win98SE it only saw the first floppy drive.

```
IMGMOUNT 0 empty -fs none -t floppy
IMGMOUNT 1 empty -fs none -t floppy -size 512,80,2,15
IMGMOUNT C hdd.img -ide 1m
IMGMOUNT D empty -t iso -ide 2m
```

Also, when I tried to mount a CD from the drop-down menus (DOS -> Change current CD image) while running Win98SE, DOSBox-X segfaulted.

Doing the same with a 3.5" diskette image did work.

**Comment by joncampbell123** [comment]
> Thanks,
> 
> That works for a 3.5" and CD-ROM drive, it does not seem to work for a 5.25". Or at least, after booting Win98SE it only saw the first floppy drive.
> 
> ```
> IMGMOUNT 0 empty -fs none -t floppy
> IMGMOUNT 1 empty -fs none -t floppy -size 512,80,2,15
> IMGMOUNT C hdd.img -ide 1m
> IMGMOUNT D empty -t iso -ide 2m
> ```
> 
> Also, when I tried to mount a CD from the drop-down menus (DOS -> Change current CD image) while running Win98SE, DOSBox-X segfaulted.
> 
> Doing the same with a 3.5" diskette image did work.

Remember that -size is sz,S,H,C not sz,C,H,S.

IMGMOUNT 1 empty -fs none -t floppy -size 512,15,2,80

That's how the original DOSBox project defined it and DOSBox-X follows that for compatibility.

**Comment by rderooy** [comment]
Thanks for the correction. But it makes no difference, Win98SE still does not show the drive B icon. Here is what the logs print for the mount commands.

```
LOG: FAT: Partition type is MBR (IBM PC)
LOG: Mounted FAT volume is FAT32 with 514573 clusters
LOG: Mapping BIOS DISK C/H/S 1023/64/63 as IDE 4092/16/63
LOG: IMGMOUNT: HDD image mounted to drive no. 2 (IDE Primary Master)
LOG: Mounted empty C/H/S/sz 80/2/18/512 1440KB
LOG: Updating floppy type to 8 BIOS type 0x02
LOG: Mounted empty C/H/S/sz 80/2/15/512 1200KB
LOG: Empty ISO
LOG: IMGMOUNT: CD-ROM image mounted to drive D (IDE Secondary Master)
```

**Comment by rderooy** [comment]
Some further testing show that the issue seems to be that Win98SE only sees the first floppy drive, regardless of configuration. If I swap the two around such that A: becomes 5.25" and B: becomes 3.5", Win98SE will only show the 5.25" A: drive.

**Comment by joncampbell123** [comment]
For whatever reason I can't get Windows 95 to see more than one floppy drive.

I also cannot get Windows 95 to talk to the floppy controller emulation. I don't mean like it pokes that the I/O ports and rejects it, I mean it absolutely won't even *try* to probe the FDC to talk to it. It doesn't *try* anything and then complains that the FDC is not present or working.

**Comment by frank-deng** [comment]
When I boot into real DOS with an empty CD-ROM drive, then load CD-ROM disc from menu "DOS->Change current CD image...", the whole emulator will **CRASH**!

Changing floppy disk from menu "DOS->Change current floppy image..." works well without crashing when booting into real DOS with empty floppy drives.

<img width="277" alt="empty_drive" src="https://github.com/joncampbell123/dosbox-x/assets/8265668/8409203d-2f7f-4cf5-8447-ab843711e1a0">

Also is it possible to add an action "Leave Empty" apart from Yes and No for this dialog? So as to emulate removing media from Floppy/CDROM drive.

**Comment by frank-deng** [comment]
> When I boot into real DOS with an empty CD-ROM drive, then load CD-ROM disc from menu "DOS->Change current CD image...", the whole emulator will **CRASH**!
> 
> Changing floppy disk from menu "DOS->Change current floppy image..." works well without crashing when booting into real DOS with empty floppy drives.
> 
> <img alt="empty_drive" width="277" src="https://private-user-images.githubusercontent.com/8265668/311445429-8409203d-2f7f-4cf5-8447-ab843711e1a0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIyODE1NzIsIm5iZiI6MTc0MjI4MTI3MiwicGF0aCI6Ii84MjY1NjY4LzMxMTQ0NTQyOS04NDA5MjAzZC0yZjdmLTRjZjUtODQ0Ny1hYjg0MzcxMWUxYTAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMxOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMThUMDcwMTEyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjlkZjA1MDU1OGI2N2NlYTI4MDk0N2I1NmNlNDQyMjQxZTU3ZWRlZjkwZGE5MTM5MzQ2NzU3YmZiMzBlZmQ5ZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.4U1DH6NoTkAWOO5KzXhaE8q8CF7Q39NQWPOHby47ZIg">
> Also is it possible to add an action "Leave Empty" apart from Yes and No for this dialog? So as to emulate removing media from Floppy/CDROM drive.

Already fixed.

**Comment by frank-deng** [comment]
Is it possible to add a feature of ejecting a loaded floppy/CD-ROM, like other emulators does, e.g. 86box?
